chia/server/ws_connection.py

```diff
@@ -391,34 +400,135 @@ async def _wait_and_retry(self, msg: Message, queue: asyncio.Queue):
+        if message.type == ProtocolMessageTypes.wrapped_compressed.value:
+            # Just return the message if already compressed
+            # (happens in case of retrying a rate limited message)
+            self.log.debug(
+                "_potentially_compress called with already compressed message, "
+                f"compressed size {len(message.data)}. OK if rate-limited before."
+            )
+            return message
+        # If the receiver can handle compressed messages,
+        # and we are configured to send them compressed,
+        # and the message is large enough to be beneficial for compression (CPU-wise and space on the wire)
+        # then compress it by transforming the original message into a new compressed message
+        if (
+            self.sending_compressed_enabled
+            and Capability.CAN_DECOMPRESS_MESSAGES in self.peer_capabilities
+            and len(message.data) >= self.compress_if_at_least_size
+        ):
+            compressed = Message(
+                uint8(ProtocolMessageTypes.wrapped_compressed.value),
+                message.id,
+                bytes([message.type]) + bytes(zstd.compress(message.data))
+                # the immediate line above is in practice a serialized WrappedCompressed
+            )
+            # Validating and safe-guarding
+            if get_decompressed_size(compressed.data[1:]) == len(message.data):
+                self.log.debug(
+                    f"Message compressed for sending: was {len(message.data)} bytes, "
+                    f"compressed to {len(compressed.data)}"
+                )
+                return compressed
+            else:
+                # If negative something is wrong with the data (eg. not singlesegment)
+                self.log.warning(
+                    f"Compression generated faulty result: #{get_decompressed_size(compressed.data[1:])}. "
+                    "Sending uncompressed instead."
+                )
+                return message
+        return message
+
     async def _send_message(self, message: Message):
         encoded: bytes = bytes(message)
         size = len(encoded)
         assert len(encoded) < (2 ** (LENGTH_BYTES * 8))
+
+        # 'message' can already be compressed, if rate-limited and now retried
+
+        message_type = ProtocolMessageTypes(message.type)
+        if message_type == ProtocolMessageTypes.wrapped_compressed:
+            if len(message.data) > 0:
+                message_type = ProtocolMessageTypes(message.data[0])
+
+        message_to_send = await self._potentially_compress(message)
+
+        if message_to_send != message:
+            encoded = bytes(message_to_send)
+            size = len(encoded)
+
         if not self.outbound_rate_limiter.process_msg_and_check(
-            message, self.local_capabilities, self.peer_capabilities
+            message_to_send, self.local_capabilities, self.peer_capabilities
```

Comment:
it would seem simpler to just pass `message` here, just like in the current logic. it would create a cleaner layering of where messages are compressed. I don't think you would need to consider the case where an already compressed message is passed to `_potentially_compress()` either

Reply:
The message_to_send is needed here to use the updated message size when determining if it should be rate limited or not
Regarding point two: If a compressed message was rate-limited it is put back on the outgoing_queue, it shouldn't be compressed once more.

Reply:
the cost of re-compressing might be low compared to the complexity of not having a clean boundary separating compressed and uncompressed messages. but it's a judgment call.